### PR TITLE
Two Pointers Add Walkthrough since It's kinda confusing

### PR DESCRIPTION
**Pull request by Aadit-Ambadkar**
Oh and I also fixed a mistake in a sentence.

### CODE REVIEW

**Review by SansPapyrus683 — APPROVED** [review]
i'm just going to trust you on this since i'm bad with html

**Comment by nathangong** [comment]
You could also just use markdown tables

**Review by brad182 — APPROVED** [review]

File: content/3_Silver/Two_Pointers.mdx [hunk 1]
```diff
@@ -44,16 +44,146 @@ of an interval. It can also be used to track two values in an array as shown in
 
 We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
 
-To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ both at the start of the segment respectively and use two pointers to track the interval.
```
Comment (HarryWangATX) [inline content/3_Silver/Two_Pointers.mdx:47]: i think the former is correct 👀

File: content/3_Silver/Two_Pointers.mdx [hunk 2]
```diff
@@ -44,16 +44,146 @@ of an interval. It can also be used to track two values in an array as shown in
...
 
 For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment is the maximum while being less than $t$.
 
 We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
 
-We can then continually increment $\texttt{left}$ by 1.
+We can then continually increment $\texttt{left}$ by 1. After incrementing the left index, the time used decreases, hence the right pointer never has to move back (left). Thus:
```
Comment (HarryWangATX) [inline content/3_Silver/Two_Pointers.mdx:53]: kind of redundant and unclear.

**Review by HarryWangATX — CHANGES_REQUESTED** [review]

File: content/3_Silver/Two_Pointers.mdx (hunk 1 above)
Comment (maggie-j-liu) [reply content/3_Silver/Two_Pointers.mdx:47]: this is for books, right? i'm pretty sure both pointers start at the beginning

but delete "respectively"

**Comment by maggie-j-liu** [comment]
> You could also just use markdown tables

i think those are harder to style

File: content/3_Silver/Two_Pointers.mdx (hunk 1 above)
Comment (HarryWangATX) [reply content/3_Silver/Two_Pointers.mdx:47]: but right marks the end of the segment not the start even though they all start at the same point at the beginning.

Comment (maggie-j-liu) [reply content/3_Silver/Two_Pointers.mdx:47]: ohhh i thought it was saying that they both start at the beginning, and the "tracks the interval" part meant that they were the left and right side of the segment

maybe could be reworded to make it more clear

Comment (HarryWangATX) [reply content/3_Silver/Two_Pointers.mdx:47]: yea i think we can reword this to be a bit clearer

Comment (Aadit-Ambadkar) [reply content/3_Silver/Two_Pointers.mdx:47]: Yeah let me reword that

File: content/3_Silver/Two_Pointers.mdx [hunk 3]
```diff
@@ -44,15 +44,145 @@ of an interval. It can also be used to track two values in an array as shown in
 
 We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
 
-To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ to **represent** the beggining and end of the **segment**. However, both will start at the beggining of the array. These numbers can be thought of as pointers, hence the name "two pointers".
```
Comment (ryanchou-dev) [inline content/3_Silver/Two_Pointers.mdx:47]:
```suggestion
To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ to **represent** the beginning and end of the **segment**. However, both will start at the beginning of the array. These numbers can be thought of as pointers, hence the name "two pointers".
```
spelling mistake

File: content/3_Silver/Two_Pointers.mdx [hunk 4]
```diff
@@ -44,15 +44,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<th>Right Pointer</th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+			<th></th>
+			<th></th>
+			<th></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+We can move the right pointer only to index $1$:
```
Comment (ryanchou-dev) [inline content/3_Silver/Two_Pointers.mdx:89]:
```suggestion
We can move the right pointer to index $1$:
```
redundant

File: content/3_Silver/Two_Pointers.mdx [hunk 5]
```diff
@@ -44,15 +44,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<th>Right Pointer</th>
+			<th></th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+			<th></th>
+			<th></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+The sum of the values in this range is 4, and there are 2 values. So, the current maximum is 2. Now, moving the left pointer up by 1, we can subtract 3 from the sum of values, to get 1. So, the current sum of values is 1. The array then looks like:
```
Comment (ryanchou-dev) [inline content/3_Silver/Two_Pointers.mdx:121]:
```suggestion
The sum of the values in this range is 4, and there are 2 values. So, the current maximum is 2. By incrementing the left pointer by 1, we can subtract 3 from the sum of values to get 1. The array then looks like:
```

File: content/3_Silver/Two_Pointers.mdx [hunk 6]
```diff
@@ -44,15 +44,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<th>Right Pointer</th>
+			<th></th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+			<th></th>
+			<th></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+Now, we can move the right pointer up by $2$, to the end. This makes the sum of values precisely $5$ (since $1$ (previous cur val) $+2+1=5$, and has a length of 3. So, the maximum is now $3$.
```
Comment (ryanchou-dev) [inline content/3_Silver/Two_Pointers.mdx:153]:
```suggestion
Now, we can move the right pointer to the end. This makes the sum of values $5$, since $1$ (previous cur val) $+2+1=5$, and has a length of 3. So, the maximum is now $3$.
```
> (since 1 (previous cur val)) +2+1=5,

1 + 2 + 1 = 5?

File: content/3_Silver/Two_Pointers.mdx [hunk 7]
```diff
@@ -44,15 +44,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<th>Right Pointer</th>
+			<th></th>
+			<th></th>
+			<th></th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+At this point we are done, since the right pointer has hit the end. So, the answer is $3$.
```
Comment (ryanchou-dev) [inline content/3_Silver/Two_Pointers.mdx:185]:
```suggestion
At this point we are done, since the right pointer has reached the end. So, the answer is $3$.
```

**Review by ryanchou-dev — CHANGES_REQUESTED** [review]
some grammar and spelling mistakes, but the diagrams are great!

**Comment by ryanchou-dev** [comment]
https://github.com/cpinitiative/usaco-guide/pull/2086#discussion_r688062865

is this a mistake? 🤔

File: content/3_Silver/Two_Pointers.mdx (hunk 6 above)
Comment (Aadit-Ambadkar) [reply content/3_Silver/Two_Pointers.mdx]: 😑

**Comment by Aadit-Ambadkar** [comment]
> [#2086 (comment)](https://github.com/cpinitiative/usaco-guide/pull/2086#discussion_r688062865)
> 
> is this a mistake? 🤔

Yes lol. Fixed it

**Comment by nathangong** [comment]
The "left pointer" header looks slightly misaligned with all the other headers in the table:
image

File: content/3_Silver/Two_Pointers.mdx [hunk 8]
```diff
@@ -44,15 +44,145 @@ of an interval. It can also be used to track two values in an array as shown in
 
 We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
 
-To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ to **represent** the beginning and end of the **segment**. However, both will start at the beginning of the array. These numbers can be thought of as pointers, hence the name "two pointers".
```
Comment (nathangong) [inline content/3_Silver/Two_Pointers.mdx:47]: Any reason why "represent" and "segment" are bolded?

File: content/3_Silver/Two_Pointers.mdx [hunk 9]
```diff
@@ -44,15 +44,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
 For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment is the maximum while being less than $t$.
 
 We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
 
-We can then continually increment $\texttt{left}$ by 1.
-
-**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**
+We can then continually increment $\texttt{left}$ by 1. After incrementing the left index, the time used decreases, hence the right pointer never has to move back (left). Thus:
+
+**Since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**
```
Comment (nathangong) [inline content/3_Silver/Two_Pointers.mdx:55]: Don't know if this needs to be bolded either

File: content/3_Silver/Two_Pointers.mdx (hunk 8 above)
Comment (Aadit-Ambadkar) [reply content/3_Silver/Two_Pointers.mdx:47]: I thought it made it more clear

File: content/3_Silver/Two_Pointers.mdx (hunk 9 above)
Comment (Aadit-Ambadkar) [reply content/3_Silver/Two_Pointers.mdx:55]: That was bolded before my commits

Comment (nathangong) [reply content/3_Silver/Two_Pointers.mdx:56]: Yeah I just don't know what the purpose the holding serves. Time complexity analysis isn't normally bolded in any other solution/explanation. But yeah it's ultimately up to ur discretion

**Comment by nathangong** [comment]
> The "left pointer" header looks slightly misaligned with all the other headers in the table:
> 
> image
> 
> 

Were you able to address this

**Comment by Aadit-Ambadkar** [comment]
> > The "left pointer" header looks slightly misaligned with all the other headers in the table:
> > image
> 
> Were you able to address this

No, I was not able to do so. I couldn't figure out why.

**Comment by maggie-j-liu** [comment]
i think headers might be automatically center aligned?

**Comment by nwatx** [comment]
> i think headers might be automatically center aligned?

this is true:

image

**Comment by Aadit-Ambadkar** [comment]
Check now?

**Comment by Aadit-Ambadkar** [comment]
Is there anything else? If not can someone merge this?

**Comment by SansPapyrus683** [comment]
think mods have to approve given that yours adds an editorial

**Comment by nathangong** [comment]
The first table looks good, but I think all the other ones still have center aligned headers

File: content/3_Silver/Two_Pointers.mdx [hunk 10]
```diff
@@ -44,15 +45,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<th>Right Pointer</th>
+			<th></th>
+			<th></th>
+			<th></th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+At this point we are done, since the right pointer has reached the end. So, the answer is $3$.
```
Comment (nathangong) [inline content/3_Silver/Two_Pointers.mdx:186]: You just used "so" as the transition word two sentences ago, so it might be better to switch it up 
```suggestion
Since the right pointer has reached the end of the array, we are done at this point. This leaves us with $3$ as our answer.
```

File: content/3_Silver/Two_Pointers.mdx [hunk 11]
```diff
@@ -44,15 +45,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<th>Right Pointer</th>
+			<th></th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+			<th></th>
+			<th></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+Now, we can move the right pointer to the end. This makes the sum of values $4$, since $1$ (previous cur val) $+2+1=4$, and has a length of 3. So, the maximum is now $3$.
```
Comment (nathangong) [inline content/3_Silver/Two_Pointers.mdx:154]:
```suggestion
Now, we can move the right pointer to the end. This makes the sum of values $4$, since $1$ (previous cur val) $+2+1=4$, and the length of the segment equal to 3. So, the maximum segment length is now $3$.
```

File: content/3_Silver/Two_Pointers.mdx [hunk 12]
```diff
@@ -44,15 +45,145 @@ of an interval. It can also be used to track two values in an array as shown in
 
 We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
 
-To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ to **represent** the beginning and end of the **segment**. However, both will start at the beginning of the array. These numbers can be thought of as pointers, hence the name "two pointers".
 
 For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment is the maximum while being less than $t$.
 
 We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
```
Comment (nathangong) [inline content/3_Silver/Two_Pointers.mdx:52]: Shoudn't this be the maximum value of right - left + 1?

File: content/3_Silver/Two_Pointers.mdx [hunk 13]
```diff
@@ -44,15 +45,145 @@ of an interval. It can also be used to track two values in an array as shown in
 
 We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
 
-To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ to **represent** the beginning and end of the **segment**. However, both will start at the beginning of the array. These numbers can be thought of as pointers, hence the name "two pointers".
 
 For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment is the maximum while being less than $t$.
```
Comment (nathangong) [inline content/3_Silver/Two_Pointers.mdx:50]:
```suggestion
For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment is maximized while being less than or equal to $t$.
```

**Review by nathangong — CHANGES_REQUESTED** [review]

File: content/3_Silver/Two_Pointers.mdx (hunk 12 above)
Comment (Aadit-Ambadkar) [reply content/3_Silver/Two_Pointers.mdx:52]: Yeah, it should.
```suggestion
We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}+1$ value.
```

**Comment by Aadit-Ambadkar** [comment]
> The first table looks good, but I think all the other ones still have center aligned headers

@nathangong should be fixed

File: content/3_Silver/Two_Pointers.mdx [hunk 14]
```diff
@@ -44,15 +45,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<td>Right Pointer</td>
+			<th></th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+			<th></th>
+			<th></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+Now, we can move the right pointer to the end. This makes the sum of values $4$, since $1$ (previous cur val) $+2+1=4$, and the length of the segment equal to 3. So, $\texttt{ans}$ is now $3$.
```
Comment (bqi343) [inline content/3_Silver/Two_Pointers.mdx:154]: `previous cur val`

what's cur val?

Comment (bqi343) [reply content/3_Silver/Two_Pointers.mdx:154]:
```suggestion
Now, we can move the right pointer to the end. This makes the sum of values $1+2+1=4$ and the length of the segment equal to 3. So, $\texttt{ans}$ is now $3$.
```

Comment (bqi343) [reply content/3_Silver/Two_Pointers.mdx:154]: not consistent about surrounding numbers with $$

**Review by bqi343 — CHANGES_REQUESTED** [review]

File: content/3_Silver/Two_Pointers.mdx [hunk 15]
```diff
@@ -44,15 +45,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
-To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.
-
-For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment is the maximum while being less than $t$.
-
-We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
-
-We can then continually increment $\texttt{left}$ by 1.
-
-**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ to **represent** the beginning and end of the **segment**. However, both will start at the beginning of the array. These numbers can be thought of as pointers, hence the name "two pointers."
```
Comment (bqi343) [inline content/3_Silver/Two_Pointers.mdx:48]: why still bolded?

File: content/3_Silver/Two_Pointers.mdx [hunk 16]
```diff
@@ -44,15 +45,145 @@ of an interval. It can also be used to track two values in an array as shown in
...
+			<td>Right Pointer</td>
+			<th></th>
+			<th className="bg-yellow-100 dark:bg-yellow-800"></th>
+			<th></th>
+			<th></th>
+		</tr>
+	</tbody>
+</table>
+</center>
+Now, we can move the right pointer to the end. This makes the sum of values $1+2+1=4$ and the length of the segment equal to 3. So, $\texttt{ans}$ is now $3$.
```
Comment (bqi343) [inline content/3_Silver/Two_Pointers.mdx:154]: still not consistent about surrounding numbers with $$

**Review by bqi343 — CHANGES_REQUESTED** [review]

**Review by nwatx — APPROVED** [review]

**Review by bqi343 — APPROVED** [review]
